File: src/main/java/bad/robot/radiate/teamcity/JsonProjectsUnmarshaller.java
package bad.robot.radiate.teamcity;

import bad.robot.http.HttpResponse;
import bad.robot.radiate.Unmarshaller;
import com.google.gson.Gson;

@Deprecated
class JsonProjectsUnmarshaller implements Unmarshaller<HttpResponse, Iterable<Project>> {

    @Override
    public Iterable<Project> unmarshall(HttpResponse response) {
        return new Gson().fromJson(new JsonResponse(response).body(), Projects.class);
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/JsonProjectsUnmarshaller.scala
package bad.robot.radiate.teamcity

import argonaut.Argonaut._
import simplehttp.HttpResponse
import bad.robot.radiate.{ParseError, Unmarshaller}
import bad.robot.radiate.Error

import scalaz.\/
import scalaz.syntax.std.either._

class JsonProjectsUnmarshaller extends Unmarshaller[HttpResponse, Iterable[Project]] {
  def unmarshall(response: HttpResponse): Error \/ Iterable[Project] = {
    val json = new JsonResponse(response).body
    json.decodeEither[Projects].disjunction.leftMap(ParseError)
  }
}